Added basic EXT_mesh_features samples

javagl opened the pull request:
This is an initial **DRAFT** of some examples for `EXT_mesh_features`.

It is not yet decided whether these samples should be here, because they are not strictly speaking "3D tiles samples". Also, there are some kinds of possible samples for `EXT_mesh_features` that cannot sensibly be part of a tileset (for example, an [empty asset that _only_ defines an example schema](https://github.com/javagl/JglTF/blob/master/jgltf-impl-v2-ext-mesh-features/data/EmptyAssetWithExtMeshFeatures.gltf)). But they may serve the purpose of getting an idea of which samples could make sense, and in how far certain functionalities are already supported in CesiumJS.

The current state of this PR is "bare", without any READMEs, screenshots or documentation. These will be added at a later point, as appropriate.

There currently are 5 assets (each wrapped into a trivial tileset) in https://github.com/javagl/3d-tiles-samples/tree/add-ext-mesh-features-samples/EXT_mesh_features . Each asset has the same structure: It consists of a single scene with a single mesh with a single primitive that just consists of 4 quads:

![Cesium Metadata Sandcastle](https://user-images.githubusercontent.com/5597569/149023106-55460be2-3fca-437e-920e-d241889b42bb.png)

The flavors of this asset, reflected in the test cases, are:

1. "ExplicitFeatureIds": Defines explicit feature IDs (without property tables)
2. "ExplicitFeatureIdsAndSimpleProperty": Defines explicit feature IDs and a simple VEC3/FLOAT32 property
3. "ImplicitFeatureIds": Defines implicit feature IDs (without property tables)
4. "ImplicitFeatureIdsAndSimpleProperty": Defines implicit feature IDs and a simple VEC3/FLOAT32 property
5. "MultipleFeatureIdsAndProperties": Defines multiple feature IDs and different properties


- Cases 1 and 3 do not work yet, because of https://github.com/CesiumGS/cesium/issues/9884

- Case 2 _seems_ to work, but the displayed values for the vectors is not correct. The vectors should have the values

               float exampleData[] = new float[] {
                    0.0f, 0.1f, 0.2f,
                    1.0f, 1.1f, 1.2f,
                    2.0f, 2.2f, 2.2f,
                    3.0f, 3.1f, 3.2f,
                };
   The values that are actually shown _look_ like there might be an issue related to endianness, but I haven't investigated it in detail yet. (I can only say that when adding an _accessor_ for these values, the right values are displayed when inspecting the accessor in the VS Code glTF plugin - so they should _probably_ be correct there)

- Case 4 does not seem to work - it does not seem to find the feature IDs when they are defined implicitly.

- Case 5 causes everything to collapse and splill out obscure error messages...

**NOTE**: Each and every error _might_ be an issue with the sample data itself ... or with CesiumJS. This has to be figured out as we move on. But particularly the last example is "relatively complex", and it's not unlikely that it is invalid in one way or another...

---

In order to quickly tests this, a first draft of the Sandcastle code (that would eventually go into the READMEs) is in https://github.com/javagl/3d-tiles-samples/blob/add-ext-mesh-features-samples/EXT_mesh_features/CommonSandcastleCode.txt 

It contains a code block
```
var testCase = "";
//testCase = "ExplicitFeatureIds";
//testCase = "ExplicitFeatureIdsAndSimpleProperty";
//testCase = "ImplicitFeatureIds";
//testCase = "ImplicitFeatureIdsAndSimpleProperty";
testCase = "MultipleFeatureIdsAndProperties";
```
that allows to quickly select the example that should be loaded, for debugging and testing.

ptrgags commented:
@javagl tried out your examples. I can reproduce the problems for the first 4. 

For the 5th one, it loads fine, though it looks like the string property is misalligned, see here how the property value is "Rai" when I click the bottom left square, and if I click the bottom right one I get the "n" for "Rain"

![image](https://user-images.githubusercontent.com/8422414/149026031-c531a65b-b495-4e67-bb76-89c7eb7b6595.png)

lilleyse commented:
One more sample idea that could be considered for a follow up PR - two primitives with features that point to the same property table.

javagl commented:
@ptrgags

> I can reproduce the problems for the first 4.

There was a wrong index for the property tables that caused the error that I originally considered to be some endianness issue. Sorry about that, it should be fixed with the last commit.

javagl commented:
Just noticed a typo in the example data...

> `2.0f, 2.2f, 2.2f,` 

That should have been 2.1. It's the details... (will update that at some point...)

javagl commented:
In response to https://github.com/CesiumGS/3d-tiles-samples/pull/40#issuecomment-1011101728 , I just added an example that contains two primitives that refer to the same property table. I'm still in the process of trying to find "the best way" to create these samples, but I'll consider and try out other possible (more "complex") examples, and add them here when ready.

javagl commented:
The `MultipleFeatureIdsAndProperties` example has been updated. Now it properly displays the strings:

![Cesium Metadata Strings](https://user-images.githubusercontent.com/5597569/150016591-48d4a580-e885-47d6-9a4b-c447b6fe602f.png)

The reason for the wrong alignment was that I had created the buffer views as `GL_ARRAY_BUFFER` buffer views. These require a `byteStride` of at least 4, and that caused the layout of the data to be changed accordingly. That's wrong when this is about a buffer view that does not have an accessor, is only used for metadata. In the latest state, they use an undefined `target` (indicating that the data is tightly packed).

javagl commented:
I have added a first, basic example for Feature ID Textures. It might look familiar to some of you. 

![Cesium FirstFeatureIdTexture](https://user-images.githubusercontent.com/5597569/150411668-e7616b58-1c58-461b-8026-eb7ca2d1bf81.png)

I'll have to look into the details on how the sandcastle has to be updated for this (i.e. to actually look up the feature ID properly - right now, the tooltip always shows the info for Feature ID 0 (Wall)). 

@ptrgags Even though texture filtering is set to 9728 (NEAREST), there seems to be some "blurring". The textures right now have a size of 10x10 - could it be that they are resized to a power-of-2 (for some WebGL 1 compatibility or so)? 

(For comparison, the screenshot shows the three.js viewer result). 

And I'll attach the textures here, just for the giggles:

![featureIdTexture](https://user-images.githubusercontent.com/5597569/150412063-9ab6bd82-2a9d-4b8b-95ea-71eeda5c3be4.png)

![featureIdTexture_color](https://user-images.githubusercontent.com/5597569/150412077-02bf09fb-fe66-49e6-b437-3ecafcd14251.png)

(Yes, I'll probably replace the color texture with a nicer (higher-resolution) one, this is just a first test...)

javagl commented:
> ... right now, the tooltip always shows the info for Feature ID 0 (Wall)).

That was actually caused by the feature ID being stored in the wrong channel. This is fixed with the last commit.

Also, changed the size of the texture to 16x16 to avoid upsampling. The actual _color_ texture is now a larger, "artificially pixelated" image, like this:

![SimpleFeatureIdTextureScreenshot](https://user-images.githubusercontent.com/5597569/150649323-e8c9dbfd-0b2d-473f-b731-30753f40a8e3.gif)

I think this makes sense, but I can also replace that texture with something that more resembles "a real house", if that is preferred.

javagl commented:
I have added some basic documentation for the first examples:

- [Documentation for `ExplicitFeatureIds`](https://github.com/javagl/3d-tiles-samples/blob/add-ext-mesh-features-samples/EXT_mesh_features/ExplicitFeatureIds/README.md) (showing the structure of the mesh primitive and the IDs)
- [Documentation for `ExplicitFeatureIdsAndSimpleProperty`](https://github.com/javagl/3d-tiles-samples/tree/add-ext-mesh-features-samples/EXT_mesh_features/ExplicitFeatureIdsAndSimpleProperty), which just _refers_ to the first one for the "common" part, and adds information about the actual property, as well as a sandcastle example

- [ ] TODO Add the sandcastle for the first example, when feature IDs without property tables work

javagl commented:
I have added a (preliminary) "complex" example that can (to some extent) be used for testing some of the changes from https://github.com/CesiumGS/cesium/pull/10018 

The example contains two classes, where the second one contains some of the "less trivial" types, namely

- a variable-length ARRAY normalized INT8 property
- a fixed-length ARRAY ENUM property

It also contains two feature IDs (and implicit and an explicit one). The https://github.com/javagl/3d-tiles-samples/blob/add-ext-mesh-features-samples/EXT_mesh_features/CommonSandcastleCode.txt has been updated with a basic toolbar box for selecting the `FEATURE_ID_0/1`, respectively. 

From a first test, it basically ~"seems to work". Some changes have to be done for the finalization:

- add a variable-length string array property (the most complex type, structurually)
- Improve the selection of the feature ID in the sandcastle (and maybe clean up the `obtain... ` functions based on the latest state from the PR)
- give it a more sensible name, maybe...
  <sub>And not only for this example, but for all of them:</sub>
- add documentation

javagl commented:
The last commit adds a "Complex Types" example, showing...

- variable-length (normalized) UINT8 arrays
- fixed-length BOOLEAN arrays
- variable-length STRING arrays
- fixed-length ENUM arrays

![Cesium Metadata complex types](https://user-images.githubusercontent.com/5597569/151461837-494f1f15-4df1-4f8e-9ae2-598d3fcec6ba.png)

I'll probably trim down the "Complex Example" that I added earlier, to omit the array types that had been used there, so that it solely serves as a dedicated sample for multiple classes.

javagl commented:
The last commit brings this into a state that could be (close to) ready for a first review.

The main change is that I added documentation, as a `README.md` in each example. Since the examples are, to some extent, "incremental" or show different "dimensions" independently, there is some potential for overlap and redundancies...:

- [ExplicitFeatureIds](https://github.com/javagl/3d-tiles-samples/tree/add-ext-mesh-features-samples/EXT_mesh_features/ExplicitFeatureIds) shows **only** explicit feature IDs (without property tables). [ImplicitFeatureIds](https://github.com/javagl/3d-tiles-samples/tree/add-ext-mesh-features-samples/EXT_mesh_features/ImplicitFeatureIds) shows the same feature IDs (without property tables), but this time, implicit. They link to each other, and contain a description of the basic mesh primitive structure.  

- [ExplicitFeatureIdsAndSimpleProperty](https://github.com/javagl/3d-tiles-samples/tree/add-ext-mesh-features-samples/EXT_mesh_features/ExplicitFeatureIdsAndSimpleProperty) and [ImplicitFeatureIdsAndSimpleProperty](https://github.com/javagl/3d-tiles-samples/tree/add-ext-mesh-features-samples/EXT_mesh_features/ImplicitFeatureIdsAndSimpleProperty) are the counterparts of those above. They both define the same metadata schema and property table (only accessed with different types of IDs). So they link to each other, and each of them links to its "no-property" counterpart (to have a pointer to the mesh primitive structure)  

- [TwoPrimitivesOnePropertyTable](https://github.com/javagl/3d-tiles-samples/tree/add-ext-mesh-features-samples/EXT_mesh_features/TwoPrimitivesOnePropertyTable) uses the same structure (schema and table) as the ones above, but assigns them to _two_ primitives. It links to `ExplicitFeatureIdsAndSimpleProperty` for the structure- and metadata definitions.

- [MultipleFeatureIdsAndProperties](https://github.com/javagl/3d-tiles-samples/tree/add-ext-mesh-features-samples/EXT_mesh_features/MultipleFeatureIdsAndProperties) defines one implicit and one explicit feature ID, so there's a pointer to the corresponding samples (in addition to the description of the "multiple properties", of course)

 - [ComplexTypes](https://github.com/javagl/3d-tiles-samples/tree/add-ext-mesh-features-samples/EXT_mesh_features/ComplexTypes) refers to `ExplicitFeatureIds` for the mesh primitive and feature ID definition, and explains the schema and tables for the "complex" types

- [MultipleClasses](https://github.com/javagl/3d-tiles-samples/tree/add-ext-mesh-features-samples/EXT_mesh_features/MultipleClasses) defines the same feature IDs as `MultipleFeatureIdsAndProperties` (and therefore, links to that), but shows how they can be used to access multiple classes

- [SimpleFeatureIdTexture](https://github.com/javagl/3d-tiles-samples/tree/add-ext-mesh-features-samples/EXT_mesh_features/SimpleFeatureIdTexture) is pretty much standalone...

- [SimplePropertyTexture](https://github.com/javagl/3d-tiles-samples/tree/add-ext-mesh-features-samples/EXT_mesh_features/SimplePropertyTexture) is also standalone, but not tested extensively.

The "Example Sandcastles" are still missing for most of them (except the `Explicit...` ones). While the "template" for them is in the [CommonSandcastleCode.txt](https://github.com/javagl/3d-tiles-samples/blob/add-ext-mesh-features-samples/EXT_mesh_features/CommonSandcastleCode.txt), this still has to be cleaned up and tailored for each specific example.

javagl commented:
The more I tried to "clean up" the `CommonSandcastleCode.txt`, the more I noticed that it was similar at least for the `EXT_mesh_features` samples. (There was some other code in it, for the other (tileset/tile/group) metadata). 

I now changed it to focus on the `EXT_mesh_features` samples, and allow selecting them at runtime:

![Cesium Metadata Sandcastle](https://user-images.githubusercontent.com/5597569/152173761-004d42f7-e5dd-4483-9dd9-de7f15f712ec.gif)

Roughly speaking, there are two options:

- Having a **minimal** sandcastle code in _each_ sample README
- Having that "common" code in its own `README.md` and then just _link_ to this README from the individual models

I'm a bit on the fence here:

- The first one causes some reduncancy (the code would largely be the same in each example, except for the model name)
- The second one adds some complexity (for setting up the selection mechanism in the sandcastle UI)

Any preferences?

lilleyse commented:
A dropdown list is more convenient for browsing the samples. And if the code is mostly the same it wouldn't be worth duplicating for each sample. So my preference is the second option.

javagl commented:
The common sandcastle code is now part of a "top-level" README - where "top-level" refers to the `EXT_mesh_features` subdirectory. (Again: These are, strictly speaking, not 3D Tiles samples...)

The "entry point" for these samples would therefore be https://github.com/javagl/3d-tiles-samples/tree/add-ext-mesh-features-samples/EXT_mesh_features

javagl commented:
I could add a first, basic "Feature ID by GPU instance" example here...

![Cesium Gpu Instance Metadata](https://user-images.githubusercontent.com/5597569/152390611-9d7a3aef-3f64-42b8-b958-36f297029a3e.png)

... but I think it's better to do this later (in a separate PR). The code that I'm using to create this has become increasingly messy, and I'd like to clean it up a bit first...

lilleyse commented on EXT_mesh_features/ComplexTypes/tileset.json:
```diff
@@ -0,0 +1,21 @@
+    },
+    "geometricError" : 1.0
+  },
+  "extensionsUsed" : [ "3DTILES_content_gltf" ],
+  "extensionsRequired" : [ "3DTILES_content_gltf" ]
```
While not strictly necessary, it would be good practice to list out the glTF extensions like in https://github.com/CesiumGS/3d-tiles/tree/main/extensions/3DTILES_content_gltf#extension-json

lilleyse commented on EXT_mesh_features/ComplexTypes/tileset.json:
```diff
@@ -0,0 +1,21 @@
+        0.0, 0.01, 0.0, 
+        0.0, 0.0, 0.5]
+    },
+    "geometricError" : 1.0
```
Usually leaf tiles have `geometricError` of `0.0`

lilleyse commented on EXT_mesh_features/ExplicitFeatureIdsAndSimpleProperty/README.md:
```diff
@@ -0,0 +1,50 @@
+
+The structure of the metadata is defined with an [`EXT_mesh_features` schema](https://github.com/CesiumGS/glTF/tree/3d-tiles-next/extensions/2.0/Vendor/EXT_mesh_features#schema-definitions) that contains a single class. The class only contains a single property, called `example_VEC3_FLOAT32`. The type of this property is a 3D vector with 32 bit floating-point components, as indicated by the type  ...
+
+### Metadata Instances
```
I believe we are calling these "entities", though I sometimes find myself gravitating towards "instances"

lilleyse reviewed (commented):
A couple cosmetic suggestions:

* The default camera view is too zoomed in, making it hard to see what's going on
* The city landscape is a bit gritty. I usually like choosing a nice grassy area as the default location for samples

The swath of examples is great, especially as debug/test data. Over time we should create more samples somewhere between debug data and real world data, but that's out of scope for this PR.

I did a high level pass over the documentation and samples and they generally looked good. Just a few minor comments from me.

javagl commented:
The inlined comments should be addressed with the last commits (each referring to one comment, applied to all README/tileset files - I hope that I didn't overlook anything).

> The default camera view is too zoomed in, making it hard to see what's going on

That also bugged me, but I thought that "whatever magic `zoomTo` does, it should be right". (As in "fit this into the viewport"). However, I saw that it's possible to add an offset there, and now, this is the initial configuration for the samples:

![Cesium Samples Orientation](https://user-images.githubusercontent.com/5597569/152648315-614e4c5c-7130-415b-b3ff-8f1197cddd61.png)

> The city landscape is a bit gritty. I usually like choosing a nice grassy area as the default location for samples

The problem (for me) with choosing a location is: There are so many of them :D So I picked the location of (what I think is) the Cesium Office. How about the fountain at the center of the Washington Square, `Cesium.Cartesian3.fromDegrees(-75.152325, 39.94704, 0)`?

![Cesium Sample Location](https://user-images.githubusercontent.com/5597569/152648623-75e536c7-da68-45b0-9926-644461dc68fd.png)

If it should be some completely neutral place outside of the city, just drop the `long/lat`, and I'll update it accordingly...

lilleyse commented:
I like the Washington Square location :smile:

javagl commented:
Depending on whether there will be a "new" set of samples (i.e. a new PR), or this one will be updated, I'd use that location then accordingly.

javagl commented:
Yes, #47 should contain everything that was contained here (the 'Feature ID by GPU instance' mentioned in https://github.com/CesiumGS/3d-tiles-samples/pull/40#issuecomment-1029196771 was not yet integrated here, but is still on an implicit TODO list)
